File: sdks/python3/blok/node/capability_manifest.py
"""Language-neutral capability manifest contract (ADR 0003)."""

from __future__ import annotations

from typing import Dict, List, NotRequired, TypedDict


class CapabilityResourceBounds(TypedDict, total=False):
    maxDurationMs: int
    maxMemoryBytes: int
    maxInputBytes: int
    maxOutputBytes: int
    maxConcurrency: int


class CapabilityManifest(TypedDict):
    version: str
    classification: str
    effects: List[str]
    capabilities: List[str]
    secrets: List[str]
    determinism: str
    idempotency: str
    maturity: str
    resources: NotRequired[CapabilityResourceBounds]
    runtimes: NotRequired[List[str]]
    triggers: NotRequired[List[str]]


def validate_capability_manifest(value: CapabilityManifest) -> CapabilityManifest:
    """Reject obviously unsafe/unsupported metadata before advertising it."""
    required = (
        "version", "classification", "effects", "capabilities", "secrets",
        "determinism", "idempotency", "maturity",
    )
    missing = [key for key in required if key not in value]
    if missing:
        raise ValueError(f"capability manifest missing: {', '.join(missing)}")
    if value["version"] != "1":
        raise ValueError("capability manifest version must be 1")
    if any("=" in secret for secret in value["secrets"]):
        raise ValueError("capability manifest secrets must be reference names, never values")
    return value

File: sdks/elixir/lib/blok/capability_manifest.ex
defmodule Blok.CapabilityManifest do
  @moduledoc "Validates the node capability boundary without atomizing wire data."

  @allowed_keys ~w(version classification effects capabilities secret_refs determinism idempotency maturity resources runtime_constraints)a

  def validate!(manifest) when is_map(manifest) do
    Enum.each(Map.keys(manifest), fn key ->
      unless is_binary(key) or key in @allowed_keys do
        raise ArgumentError, "capability manifest keys must be strings or declared atoms"
      end
    end)

    secret_refs = Map.get(manifest, "secret_refs", Map.get(manifest, :secret_refs, []))

    unless is_list(secret_refs) and Enum.all?(secret_refs, &(is_binary(&1) and &1 != "")) do
      raise ArgumentError, "capability manifest secret_refs must contain opaque reference names"
    end

    manifest
    |> Blok.Schema.to_json()
    |> Map.put_new("version", "1")
  end

  def validate!(_), do: raise(ArgumentError, "capability manifest must be a map")
end
